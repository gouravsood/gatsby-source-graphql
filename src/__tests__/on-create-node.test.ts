import onCreateNode from '../on-create-node'

// Cast fetch to a mockFn to avoid casting everytime below
const fetch: jest.Mock = require('node-fetch').default

jest.mock('node-fetch', () => ({
  default: jest.fn()
}))

const createGatsby = ({
  createNode = jest.fn(),
  createParentChildLink = jest.fn(),
  loadNodeContent = jest.fn(),
  node = {},
} = {}) => ({
  boundActionCreators: {
    createNode,
    createParentChildLink,
  },
  loadNodeContent,
  node: {
    extension: 'graphql',
    name: 'github',
    ...node,
  },
})

const createOptions = ({
  headers = {},
  queries = [],
  url = '//api.local',
  variables = {},
} = {}) => ({
  headers,
  queries,
  url,
  variables,
})

const mockFetch = ({ message = '', statusText = '' } = {}) => {
  fetch.mockReset()
  fetch.mockReturnValue({
    json: jest.fn(() => ({
      data: { id: 1 },
      message,
    })),
    ok: message.length === 0,
    statusText,
  })
}

describe('onCreateNode', () => {
  it('creates a node and links to parent', async () => {
    mockFetch()

    const createNode = jest.fn()
    const loadNodeContent = jest.fn(() => 'query { id }')
    const gatsby = createGatsby({ createNode, loadNodeContent })
    const options = createOptions()
    const actual = await onCreateNode(gatsby, options)
    const contentDigest = 'd2ce28b9a7fd7e4407e2b0fd499b7fe4'

    expect(createNode.mock.calls[0][0]).toEqual({
      children: [],
      id: `__graphql__${contentDigest}`,
      internal: {
        content: JSON.stringify({ id: 1 }),
        contentDigest,
        type: 'GithubGraphQL',
      },
    })
  })

  it('allow the use to a transform function', async () => {
    mockFetch()

    const createNode = jest.fn()
    const loadNodeContent = jest.fn(() => 'query { id }')
    const gatsby = createGatsby({ createNode, loadNodeContent })
    const options = { ...createOptions(), transform: data => ({ ...data, type: 'CustomType' }) }
    const actual = await onCreateNode(gatsby, options)
    const contentDigest = 'b035be5e1e4d06e6faaf0fc2d9f2f77f'

    expect(createNode.mock.calls[0][0]).toEqual({
      type: 'CustomType',
      id: `__graphql__${contentDigest}`,
      children: [],
      internal: {
        content: JSON.stringify({ id: 1, type: 'CustomType' }),
        contentDigest,
        type: 'CustomType',
      },
    })
  })

  it('allows Promise based options', async () => {
    mockFetch()

    const headers = {
      authorization: 'Bearer token',
    }
    const loadNodeContent = jest.fn(() => 'query { id }')
    const gatsby = createGatsby({ loadNodeContent })
    const options = Promise.resolve(createOptions({ headers }))
    const actual = await onCreateNode(gatsby, options)

    expect(fetch.mock.calls[0][1]).toMatchSnapshot()
  })

  it('allows passing custom headers', async () => {
    mockFetch()

    const headers = {
      authorization: 'Bearer token',
    }
    const loadNodeContent = jest.fn(() => 'query { id }')
    const gatsby = createGatsby({ loadNodeContent })
    const options = createOptions({ headers })
    const actual = await onCreateNode(gatsby, options)

    expect(fetch.mock.calls[0][1]).toMatchSnapshot()
  })

  it('throws an error with bad api response', async () => {
    mockFetch({ message: 'Post is malformed.', statusText: 'Bad Request' })

    const gatsby = createGatsby()
    const options = createOptions()
    const actual = onCreateNode(gatsby, options)

    await expect(actual).rejects.toThrowErrorMatchingSnapshot()
  })

  it('throws an error when null `headers`', async () => {
    const gatsby = createGatsby()
    const options = createOptions({ headers: null })
    const actual = onCreateNode(gatsby, options)

    await expect(actual).rejects.toThrowErrorMatchingSnapshot()
  })

  it('throws an error with missing or empty `url`', async () => {
    const gatsby = createGatsby()
    const options = createOptions({ url: '' })
    const actual = onCreateNode(gatsby, options)

    await expect(actual).rejects.toThrowErrorMatchingSnapshot()
  })

  it('throws an error when null `variables`', async () => {
    const gatsby = createGatsby()
    const options = createOptions({ variables: null })
    const actual = onCreateNode(gatsby, options)

    await expect(actual).rejects.toThrowErrorMatchingSnapshot()
  })

  it('throws an error when non-function `transform`', async () => {
    const gatsby = createGatsby()
    const options = { ...createOptions(), transform: 'non-function' }
    const actual = onCreateNode(gatsby, options)

    await expect(actual).rejects.toThrowErrorMatchingSnapshot()
  })

  it('does nothing when node.extensions !== `graphql`', async () => {
    const gatsby = createGatsby({ node: { extension: '' } })
    const options = createOptions()
    const actual = await onCreateNode(gatsby, options)

    expect(actual).toBeUndefined()
  })
})
