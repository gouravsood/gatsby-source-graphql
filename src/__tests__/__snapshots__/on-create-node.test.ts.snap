// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onCreateNode allows Promise based options 1`] = `
Object {
  "body": "{\\"query\\":\\"query { id }\\",\\"variables\\":{}}",
  "headers": Object {
    "authorization": "Bearer token",
    "content-type": "application/json",
  },
  "method": "post",
}
`;

exports[`onCreateNode allows passing custom headers 1`] = `
Object {
  "body": "{\\"query\\":\\"query { id }\\",\\"variables\\":{}}",
  "headers": Object {
    "authorization": "Bearer token",
    "content-type": "application/json",
  },
  "method": "post",
}
`;

exports[`onCreateNode throws an error when non-function \`transform\` 1`] = `"[Bad Request]: Post is malformed."`;

exports[`onCreateNode throws an error when null \`headers\` 1`] = `"Must supply \`headers\` option that is an object."`;

exports[`onCreateNode throws an error when null \`variables\` 1`] = `"Must supply \`variables\` option that is an object."`;

exports[`onCreateNode throws an error with bad api response 1`] = `"[Bad Request]: Post is malformed."`;

exports[`onCreateNode throws an error with missing or empty \`url\` 1`] = `"Must supply \`url\` option to point to your GraphQL API."`;
